Update docs + tweak top-level API (#18)
DataEngineeringLabs/arrow2-derive: README.md

# Arrow2-derive - derive for Arrow2

This crate enables converting between arrays of rust structures and the Arrow memory format as represented by arrow2 data structures. Specifically, it exposes a `ArrowStruct` derive macro, which can be used to annotate a structure to derive the following:
- A arrow2::array::MutableArray, which is used for serialization, and converted to an arrow2::array::Array for use by the the rest of the Arrow/arrow2 ecosystem.
- A typed arrow2::array::Array which can be used to deserialize back to the original array.

The following features are supported:

- arrow2 primitive types can be used as struct fields.
    - numeric types: [`u8`], [`u16`], [`u32`], [`u64`], [`i8`], [`i16`], [`i32`], [`i64`], [`f32`], [`f64`]

>     - other types: [`bool`], [`String`]

Reviewer: ```suggestion
    - other types: [`bool`], [`String`], [`Binary`]
```

To achieve this, the following approach is used. 

- Introduce three traits, `ArrowField`, `ArrowSerialize`, and `ArrowDeserialize` that can be implemented by types that can be represented in Arrow. Implementations are provided for the built-in arrow2` types, and custom implementations can be provided for other types.

- Blanket implementations are provided for types that recursively contain types that implement the above traits eg. [`Option<T>`], [`Vec<T>`], [`Vec<Option<T>>`], [`Vec<Vec<Option<T>>>`], etc. The blanket implementation needs be enabled by the `arrow_enable_vec_for_type` macro on the primitive type. This explicit enabling is needed since Vec<u8> is a special type in Arrow, and implementation specialization is not yet supported in rust to allow blanket implementations to coexist with more specialized implementations.

- Some additional supporting traits such as `ArrowMutableArrayTryPushGeneric` are used to workaround the unavailability of features like generic associated types.

### Why not serde?

> While, serde is the de-facto serialization framework for rust it introduces a layer of indirection that seemed unnecessary for the arrow2 framework. Specifically, arrow2 itself serves as an intermediate representation, which can be serialized/deserialized to a wide variety of formats including JSON, avro, etc, so it seemed conceptually more straightforward to have the derive functionality to have much tighter coupling with arrow2.

Reviewer: ```suggestion
While serde is the de-facto serialization framework in Rust, it introduces a layer of indirection.
Specifically, arrow2 uses Apache Arrow's in-memory columnar format, while serde is row based.
Using serde as an intermediary would lead to an extra allocation per non-primitive type per item, which is very expensive.

Arrow's in-memory format can be serialized/deserialized to a wide variety of formats including Apache Parquet, JSON, Apache Avro, Arrow IPC, and Arrow FFI specification.
```

> Also, one of the objectives of this crate is to derive a compile-time schema for rust structs, which requires an orthogonal approach to serde. This is achieved by the `ArrowField` trait. Other crates that integrate serde for example [serde_arrow](https://github.com/chmp/serde_arrow), either need an explicit schema or need to infer the schema at runtime.

Reviewer: ```suggestion
One of the objectives of this crate is to derive a compile-time Arrow schema for Rust structs, which we achieve via the `ArrowField` trait.
Other crates that integrate serde for example [serde_arrow](https://github.com/chmp/serde_arrow), 
either need an explicit schema or need to infer the schema at runtime.
```

Lastly, the serde ecosystem comes with its own default representations for temporal times, that differ from the default representations of arrow2. It seemed best to avoid any conflicts by introducing a new set of traits.

> Probably, the biggest disadvantage of not using Serde is types in codebases which already implement serde traits, will need to additionally reimplement the traits needed by this crate. If this starts to become an issue, then we can look into introducing a serde adapter to leverage those implementations.

Reviewer: ```suggestion
The biggest disadvantage of not using Serde is for types in codebases that already implement serde traits.
They will need to additionally reimplement the traits needed by this crate.
If this starts to become an issue, then we can look into introducing a serde adapter to leverage those implementations.
```
